File: _includes/code/typescript/config-refs.datatypes.date.ts
import assert from 'assert';
import weaviate, { WeaviateClient } from 'weaviate-client';

// START ConfigureDataType
import { dataType } from 'weaviate-client';

// END ConfigureDataType

const client: WeaviateClient = await weaviate.connectToLocal();

await client.collections.delete('ConcertTour');

// START ConfigureDataType
// Create collection
const myCollection = await client.collections.create({
  name: 'ConcertTour',
  properties: [
    {
      name: 'artist',
      dataType: dataType.TEXT,
    },
    {
      name: 'tour_name',
      dataType: dataType.TEXT,
    },
    {
      name: 'tour_start',
      dataType: dataType.DATE,
    },
    {
      name: 'tour_dates',
      dataType: dataType.DATE_ARRAY,
    },
  ],
  // Other properties are omitted for brevity
});
// END ConfigureDataType

// START AddObject
const exampleObject = {
  name: 'Taylor Swift',
  tour_name: 'Eras Tour',
  tour_start: new Date(2023, 3, 17),
  // Use JavaScript Date object
  tour_dates: [
    new Date(2023, 3, 17),
    new Date(2023, 3, 18),
    // .. more dates
    new Date(2024, 12, 6),
    new Date(2024, 12, 7),
  ],
  // // Or, use RFC3339 string
  // tour_dates: [
  //   '2023-03-17T00:00:00Z',
  //   '2023-03-18T00:00:00Z',
  //   // .. more dates
  //   '2024-12-07T00:00:00Z',
  //   '2024-12-08T00:00:00Z',
  // ]
};

const obj_uuid = await myCollection.data.insert(exampleObject);
// END AddObject

const returnedObject = await myCollection.query.fetchObjectById(obj_uuid);

assert.notEqual(obj_uuid, null);
assert.deepEqual(
  returnedObject?.properties.tour_start,
  exampleObject.tour_start
);

client.close();

File: _includes/code/python/config-refs.datatypes.date.py
import weaviate

# START ConfigureDataType
from weaviate.classes.config import Property, DataType
from datetime import datetime, timezone

# END ConfigureDataType

client = weaviate.connect_to_local()

# Delete collection if it exists
client.collections.delete(name="ConcertTour")

# START ConfigureDataType
# Create collection
my_collection = client.collections.create(
    name="ConcertTour",
    properties=[
        Property(name="artist", data_type=DataType.TEXT),
        Property(name="tour_name", data_type=DataType.TEXT),
        Property(name="tour_start", data_type=DataType.DATE),
        Property(name="tour_dates", data_type=DataType.DATE_ARRAY),
    ],
    # Other properties are omitted for brevity
)
# END ConfigureDataType

# START AddObject
# Create an object
# In Python, you can use the RFC 3339 format or a datetime object (preferably with a timezone)
example_object = {
    "artist": "Taylor Swift",
    "tour_name": "Eras Tour",
    "tour_start": datetime(2023, 3, 17).replace(tzinfo=timezone.utc),
    "tour_dates": [
        # Use `datetime` objects with a timezone
        datetime(2023, 3, 17).replace(tzinfo=timezone.utc),
        datetime(2023, 3, 18).replace(tzinfo=timezone.utc),
        # .. more dates
        # Or use RFC 3339 format
        "2024-12-07T00:00:00Z",
        "2024-12-08T00:00:00Z",
    ],
}

obj_uuid = my_collection.data.insert(example_object)
# END AddObject

returned_obj = my_collection.query.fetch_object_by_id(obj_uuid)

assert obj_uuid is not None
assert returned_obj.properties["tour_start"] == example_object["tour_start"]
# Only testing the `tour_start` date for simplicity (the `tour_dates` array has mixed date formats)

client.close()
